Fix R parsing HTML table with spaces, using pandoc

jackkamm opened the pull request:
Sorry to be revisiting this so soon -- I found a bug already in `jupyter-R.el` :-/

If a table has spaces, it gets parsed incorrectly, like so:
```
#+begin_src jupyter-R :session R
  data.frame(x=c("a", "b c", "d"), y=1:3)
#+end_src

#+RESULTS:
|   | x | y |
|---+---+---|
| a | 1 |   |
| b | c | 2 |
| d | 3 |   |
```
The second entry of the "x" column should be "b c" but it gets split into an extra column, and the rest of the table is mangled as a consequence.

I decided to try an alternate approach. If `pandoc.el` is available, use it to convert HTML output to org-mode. Otherwise, if libxml is available, use shr to render it -- don't try to do any HTML parsing ourselves. This simplifies the code a bit, hopefully making it more maintainable.

Here's some examples. Table output when pandoc is available:
```
#+begin_src jupyter-R :session R
  data.frame(x=c("a", "b c", "d"), y=1:3)
#+end_src

#+RESULTS:
:RESULTS:
| x     | y   |
|-------+-----|
| a     | 1   |
| b c   | 2   |
| d     | 3   |
:END:
```
Libxml table output:
```
#+begin_src jupyter-R :session R
  data.frame(x=c("a", "b c", "d"), y=1:3)
#+end_src

#+RESULTS:
:   x  y    
:   a   1     
:   b c  2     
:   d   3     
```
Pandoc output of structured list:
```
#+begin_src jupyter-R :session R :display html
  foo = data.frame(a=letters[1:3], b=1:3)
  list(foo, 1:3, c=foo)
#+end_src

#+RESULTS:
:RESULTS:
- [[1]] :: | a   | b   |
  |-----+-----|
  | a   | 1   |
  | b   | 2   |
  | c   | 3   |

- [[2]] :: 

  1. 1
  2. 2
  3. 3

- $c :: | a   | b   |
  |-----+-----|
  | a   | 1   |
  | b   | 2   |
  | c   | 3   |
:END:
```
The nested list rendering is not perfect -- the inner table headers are shifted, and [[1]] and [[2]] get rendered as links. However it's still readable and not a huge deal IMO -- can always use `:display plain` for these sorts of edge cases.

An alternative which could be cool (if you're open to it) would be to support `pandoc.el` more generally, perhaps through a `:results pandoc` header property. Then we could remove the special handling of R html results. It might also be nice for pandas dataframes (https://github.com/dzop/emacs-jupyter/issues/88) -- that way people don't have to mess with their ipython startup files.

dzop commented:
I think it makes sense to rely on `pandoc` for processing these kinds of cases. The `nbconvert` program of Jupyter already uses pandoc so its most likely available for users. There are most likely other instances where a kernel emits HTML that could be transformed into org syntax. I've also been experimenting with converting notebooks to org-mode documents using pandoc (https://github.com/dzop/emacs-jupyter/tree/feature/notebook-conversion) so I would have probably ended up doing something similar. I don't think we need to rely on `pandoc.el` though, we can just roll our own function:

```emacs-lisp
(defun jupyter-pandoc-convert (from to from-string &optional callback)
  "Use pandoc to convert a string in FROM format to TO format.
Starts a process and converts FROM-STRING, assumed to be in FROM
format, to a string in TO format and returns the converted
string.

If CALLBACK is specified, return the process object. When the
process exits, call CALLBACK with zero arguments and with the
buffer containing the converted string current."
  (cl-assert (executable-find "pandoc"))
  (let* ((process-connection-type nil)
         (proc (start-process
                "jupyter-pandoc"
                (generate-new-buffer " *jupyter-pandoc*")
                "pandoc" "-f" from "-t" to "--")))
    (set-process-sentinel
     proc (lambda (proc _)
            (when (memq (process-status proc) '(exit signal))
              (with-current-buffer (process-buffer proc)
                (funcall callback)
                (kill-buffer (process-buffer proc))))))
    (process-send-string proc from-string)
    (process-send-eof proc)
    (if callback proc
      (let ((to-string ""))
        (setq callback (lambda () (setq to-string (buffer-string))))
        (while (zerop (length to-string))
          (accept-process-output nil 1))
        to-string))))
```

Also, I forgot to mention this last time, but could you ensure that your changes byte compile cleanly by addressing the warnings you get when you run `byte-compile-file` on `jupyter-R.el`.

jackkamm commented:
I've run `byte-compile-file` and addressed the warnings, sorry about that.

Glad you're supportive of using pandoc to render HTML more generally. Therefore, I think it makes sense to remove the special pandoc/shr handling from `jupyter-R.el` entirely, which I've done.

I added general pandoc rendering for HTML in a standalone commit. There's a couple ways to go about this, and I'm not sure I picked the way you'd like, so we can reset and pursue in another branch if you prefer that.

Specifically, I added a `:results pandoc` option for HTML output, which calls out to your `jupyter-pandoc-convert`, plus a little code I added to render tables specially. Below are a couple alternative possibilities...what do you think?
1. Get rid of `:results pandoc`, and always render HTML with pandoc by default. Maybe add a `:results raw` option to render raw HTML, or do some special checking to make sure the HTML is something we'd actually want to convert (e.g. list, table, etc.)
2. At the other extreme, keep the `:results pandoc`, and also provide this option to latex and markdown outputs as well. For example, I did notice slight differences between the pandoc conversions of Markdown and HTML outputs on nested R lists. However, this is an edge case and the differences were minor.

jackkamm commented on jupyter-org-client.el:
```diff
@@ -991,9 +1021,18 @@ parsed, wrap DATA in a minipage environment and return it."
                                   &optional _metadata)
-  (jupyter-org-export-block "html" data))
+  (if (member "pandoc" (alist-get :result-params params))
+      (jupyter-org-scalar
```
On further testing, `jupyter-org-scalar` seems to append an extra line if `data` is a short string (but not if it's a long string wrapped in a quote block). It seems to be necessary when data is a table though. I'm slightly confused when I should be calling `jupyter-org-scalar`...

dzop replied:
Yes `jupyter-org-scalar` should only be passed an unformatted string or table. It is not intended to handle `org` formatted strings. If you want to handle tables specially you should wrap the `org-tabel-to-lisp` call with `jupyter-org-scalar` and the `buffer-string` call with `jupyter-org-raw-string`.

Couldn't you avoid doing all of the table processing by just doing something like

```emacs-lisp
(jupyter-org-raw-string (jupyter-pandoc-convert "html" "org" data))
```

jackkamm replied:
Reason I was using table parsing and `jupyter-org-scalar`, is that `jupyter-org-raw-string` inserts the table in a `:RESULTS:` drawer, and I was trying to replicate the behavior of `jupyter-python` and `ob-python` which doesn't have a drawer. However, I think you're right, it's better to be simple, so I'll switch to `jupyter-org-raw-string` without the table parsing.

Longer term, it might be nice to have finer-grained control over whether the table should go in a drawer. For example, I prefer a drawer for big tables but not small ones. But this is not a priority for me, and should probably be a separate issue.

dzop replied:
I think you can get around wrapping the table with a drawer by also putting an `org-table` text property on the 1st character of the table string. The result won't be wrapped in a drawer if it can be removed by `org-babel` and determining if a result needs to be wrapped is mainly handled by the function `jupyter-org-babel-result-p`.

It would probably be good to create a new function to add this text property similar to what `jupyter-org-raw-string` and `jupyter-org-raw-string-p` do but for the `org-table` property. When doing this we will have to replace all instances of the `org-table` property with the corresponding functions.

dzop commented on jupyter-org-client.el:
```diff
@@ -845,6 +845,36 @@ the width or height of the image."
         metadata
       (jupyter-org-image-link file width height))))
 
+(defun jupyter-pandoc-convert (from to from-string &optional callback)
```
Could you move this to `jupyter-mime.el` since it could be useful in other contexts in the future. You can place it at the end of the file with a new comment heading:

```emacs-lisp
;;; Pandoc
```

dzop commented:
I would prefer your option (1) to option (2), but I don't think we should do any checking of the HTML since there is most likely some quirks with how the various kernel languages emit HTML that we would have to take into account. We can revisit this if needed in the future.

I think `:results raw` should have the opposite behavior, i.e. specifying `:results raw` should use `pandoc` to convert the HTML to org syntax since that is closer to the intended meaning of `:results raw` (see `org-babel-insert-result`). I would like to keep the meaning of the header arguments as consistent as possible with how `org-mode` itself uses them to avoid any confusion. At the very least I would prefer not to add any new options to `:results`.

Another alternative would be to:

3. Provide a new header argument, `:pandoc t`, in combination with a customizable variable `jupyter-org-pandoc-convertable` (or a better name) which is a list of mime types that are convertable to org syntax using pandoc. Specifying `:pandoc t` on a code block means that if the kernel emits a mime type in `jupyter-org-pandoc-convertable`, then the associated data is converted into org syntax using pandoc. 

(3) seems like a better approach than (1) or (2) as it leaves it up to the user to specify which mime types are convertable using pandoc. What do you think?

Also I think it would be a good idea to close this pull request and open up two new ones. One with the pandoc feature and another with the changes to `jupyter-R.el`.

jackkamm commented:
(3) sounds great to me. Closing this to split into 2 separate PRs which I'll submit shortly.
